node/src/peer_tracker.rs

```diff
@@ -1,47 +1,64 @@
+use std::borrow::Borrow;
+
+use dashmap::DashSet;
 use libp2p::PeerId;
-use std::collections::HashSet;
-use std::sync::RwLock;
 
 pub struct PeerTracker {
-    inner: RwLock<Inner>,
-}
-
-struct Inner {
-    peers: HashSet<PeerId>,
+    peers: DashSet<PeerId>,
 }
 
 impl PeerTracker {
     pub fn new() -> Self {
         PeerTracker {
-            inner: RwLock::new(Inner {
-                peers: HashSet::new(),
-            }),
+            peers: DashSet::new(),
         }
     }
 
     pub fn add(&self, peer: PeerId) {
-        let mut inner = self.inner.write().unwrap();
-        inner.peers.insert(peer);
+        self.peers.insert(peer);
+    }
+
+    pub fn add_many<I, P>(&self, peers: I)
+    where
+        I: IntoIterator<Item = P>,
+        P: Borrow<PeerId>,
+    {
+        for peer in peers {
+            self.add(*peer.borrow());
+        }
     }
 
     pub fn remove(&self, peer: PeerId) {
-        let mut inner = self.inner.write().unwrap();
-        inner.peers.remove(&peer);
+        self.peers.remove(&peer);
     }
 
     pub fn len(&self) -> usize {
-        let inner = self.inner.read().unwrap();
-        inner.peers.len()
+        self.peers.len()
     }
 
     pub fn is_empty(&self) -> bool {
-        self.len() == 0
+        self.peers.is_empty()
     }
 
     pub fn best_peer(&self) -> Option<PeerId> {
         // TODO: Implement peer score and return the best.
-        let inner = self.inner.read().unwrap();
-        inner.peers.iter().next().copied()
+        self.peers.iter().next().map(|v| v.key().to_owned())
+    }
+
+    pub fn best_n_peers(&self, limit: usize) -> Option<Vec<PeerId>> {
```

Comment:
Having `Option<Vec>`  only to never return empty vec feels a bit weird to me, empty is empty :P 
Not feeling particularly strongly about it though

Reply:
Well, it works nicely with let-else:

```rust

        let Some(peers) = self.peer_tracker.best_n_peers(MAX_PEERS) else {
            respond_to.maybe_send_err(P2pError::NoPeers);
            return;
        };
```

Reply:
I was thinking about vec destructuring, but indeed there doesn't seem to be equivalentely concise syntax to `let .. else`.
`PeerTracker` seems to be part of public API? My worry is that having `Option<Vec>` suggests there's a difference between no-result and empty-result. We can weight it against convenience and I think I can live with both :P

Reply:
I don't have strong opinion tbh. I can change it to just Vec.

Reply:
me neither, I like the concise usage (which is also a part of public api, so if we want users to use it like that then maybe add some docstring with this exact example and we good?

Reply:
I prefer a simple Vec instead. I will change it later